### PR TITLE
store only removes from read cache if slot is possibly present

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
read only cache has some lock holding problems affecting performance.
One specific area is that to be safe, anytime we store an account, we make sure the account has not been added to the read only cache in that slot. This check gets a write lock on the shard in the `DashMap`. The account should never be present in the read only cache at that slot. But, it is safest to check.

#### Summary of Changes
Keep track of the highest slot ever written to the read only cache. If the slot we're writing to has never been written, then there is no reason to remove first.

Fixes #

### CODE REVIEW

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -6470,10 +6470,15 @@ impl AccountsDb {
     ) -> Vec<AccountInfo> {
         let mut calc_stored_meta_time = Measure::start("calc_stored_meta");
         let slot = accounts.target_slot();
+        if self
+            .read_only_accounts_cache
+            .can_slot_be_in_cache(accounts.target_slot())
+        {
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs:6476]: In `write_ancient_accounts()`, we will call this function too

https://github.com/anza-xyz/agave/blob/9cd90751f0105665328691b2cf72d2a8310b09ed/accounts-db/src/ancient_append_vecs.rs#L437

With ancients, the target slot will be old, and `can_slot_be_in_cache()` should return true. In this case, do we want to handle anything differently? Maybe your idea to take a read lock and check if the account is in the read cache first, before getting a write lock and removing?

File: accounts-db/src/read_only_accounts_cache.rs [hunk 2]
```diff
@@ -169,7 +172,15 @@ impl ReadOnlyAccountsCache {
...
     }
 
+    pub fn can_slot_be_in_cache(&self, slot: Slot) -> bool {
+        self.highest_slot_stored.load(Ordering::Acquire) >= slot
+    }
+
     pub(crate) fn remove(&self, pubkey: Pubkey, slot: Slot) -> Option<AccountSharedData> {
+        if self.highest_slot_stored.load(Ordering::Acquire) < slot {
+            return None;
+        }
```
Comment (brooksprumo) [inline accounts-db/src/read_only_accounts_cache.rs]: This makes me a little uneasy. Now we'll need to know that calls to `remove` only _maybe_ remove. Can `remove` remain unchanged? I'd prefer to update callers, as there are only a few.

Also, we could create a new function like `maybe_remove()` that wraps this logic.

File: accounts-db/src/read_only_accounts_cache.rs [hunk 3]
```diff
@@ -134,6 +136,7 @@ impl ReadOnlyAccountsCache {
     }
 
     pub(crate) fn store(&self, pubkey: Pubkey, slot: Slot, account: AccountSharedData) {
+        self.highest_slot_stored.fetch_max(slot, Ordering::Release);
```
Comment (brooksprumo) [inline accounts-db/src/read_only_accounts_cache.rs:139]: Maybe his should be moved to the bottom of the function (after updates to the cache, but before the stats update)?

If you squint, a `store(Release)` is similar to an "unlock" operation, and `load(Acquire)` is similar to a "lock" operation. And you'd wait to unlock until after performing all modifications/accesses. Hence moving the `fetch_max` to after updating the cache/queue.

For this case, I still need to think through this more for what happens when multiple threads are calling this function at the same time.

**Review by brooksprumo — COMMENTED** [review]
I'm still reasoning through this when multiple threads all call `store()` at the same time. Here's some initial comments:

File: accounts-db/src/read_only_accounts_cache.rs (hunk 2 above)
Comment (jeffwashington) [reply accounts-db/src/read_only_accounts_cache.rs]: I got rid of the changes in `remove`. Those were left over from previous attempts at getting this right.

File: accounts-db/src/read_only_accounts_cache.rs (hunk 3 above)
Comment (jeffwashington) [reply accounts-db/src/read_only_accounts_cache.rs:139]: I think I'd rather it be at beginning. We can afford to be very conservative here. Once we try to store at `slot`, then that becomes the highest slot we've attempted to store at.

Comment (brooksprumo) [reply accounts-db/src/read_only_accounts_cache.rs:139]: Gotcha. Yeah, I'm making sure to think through any implications on correctness. IIUC, having the fetch_max at the top is a perf benefit, so subsequent/concurrent calls to `store()` may be able to skip removes.

Comment (jeffwashington) [reply accounts-db/src/read_only_accounts_cache.rs:139]: having at top means before we ever insert into cache, we note that we have written to this slot. That way when we check if we could have written to the cache, we will say we could have even before we actually have. This leads to more conservative behavior and avoids a race condition.

File: accounts-db/src/accounts_db.rs (hunk 1 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs:6476]: i did modify this to add `remove_assume_not_present`. Thoughts?

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs:6476]: I like it. Thanks!

File: accounts-db/src/read_only_accounts_cache.rs (hunk 3 above)
Comment (brooksprumo) [reply accounts-db/src/read_only_accounts_cache.rs:139]: Ah, right, duh. I had the logic inverted in my head... Thanks for explaining.

**Review by brooksprumo — APPROVED** [review]
:shipit:

**Comment by jeffwashington** [comment]
@brooksprumo does fetch_max ordering need to be something else:
https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.fetch_max

**Comment by brooksprumo** [comment]
> @brooksprumo does fetch_max ordering need to be something else: https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.fetch_max

I think we want `Release`. Since we're not using the return value at all from `fetch_max`, we don't need to worry about the load-side of the atomic op.

**Pull request merged**